Add ModelExporter class and add preliminary support for  exporting URDF from iDynTree::Model

traversaro opened the pull request:
Add `iDynTree::ModelExporter` class and related tests. 

Currently the only format supported for export is the URDF format.

Currently the model exporter only exports a subset of the features supported in iDynTree::Model. In particular, the following features are **not supported**:
* Additional frames
* Visual and Collision Geometries
* Sensors
* Joint limits, friction parameters
* Automatically change link frames location for link whose origin does not lay on the rotation/translation axis of the parent joint, as requested by URDF
* Wrapping the ModelExporter in bindings

The PR is composed by the following two commits:  
* https://github.com/robotology/idyntree/commit/59821242669598d6e6f2578ed3e5e70b1d6e533e : Add [fpconv](https://github.com/night-shift/fpconv) as a MIT-licensed vendored dependency, to support double - shortest string exact conversion, that is currently not supported in the STL. This will be removed as soon as we can depend on [C++17's `std::to_chars`](https://en.cppreference.com/w/cpp/utility/to_chars) 
* https://github.com/robotology/idyntree/pull/554/commits/047ef63cfdd8cf78db5e6d5bf4884f4977bcc320 : Add ModelExporter class, support for exporting URDF, and tests.

My idea is to implement most of the missing features before 0.12 release, but as the limitations are currently documented, I think it make sense to first review this initial PR, and then the other features can be added in future PRs. 

This PR is based on the old closed PR https://github.com/robotology/idyntree/pull/382 .

prashanthr05 commented on src/model_io/urdf/include/private/URDFModelExport.h:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @warning This function does not support exporting sensor or solid shapes at the moment.
+ *
+ * @param[in] options the URDFParserOptions struct of options passed to the parser
```
It'd be great to document also the output and default parameters here. Also, the location in which the urdf fille will be saved.

traversaro replied:
For the default options, the user is referred to the iDynTree::ModelExportOptions, that contains docs related to the available and the default options. I will make it clear in the commit to address reviews, let me know if it is enough.

prashanthr05 commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,295 @@
+{
+
+/**
+ * Add a <origin> URDF element to the specified parent element with the specified transform.
```
if it's not an over-kill, it will be nice to document how this might look in an urdf file.

traversaro replied:
You mean like adding an example of a origin tag, i.e.: 
~~~xml
<origin xyz="0 0 0.5" rpy="0 0 0"/>
~~~
?

prashanthr05 replied:
Something like that. It might be helpful for the developers to have a visual example of code in the documentation. Because sometimes, just looking at the arguments of the API can be ambiguous.

traversaro replied:
Ack, I will do it. Consider anyhow that this function is not included in any public API, and so it is not directed to iDynTree users, but rather to people that will edit the exporter itself.

prashanthr05 replied:
Yes, I realized that. :)

fjandrad replied:
I had the same feeling when going through it, makes it even easier to review if you don't remember exactly at the moment the expected urdf format.

prashanthr05 commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,295 @@
+}
+
+/**
+ * Add a <joint> URDF element to the specified parent element with the specified joint info.
```
We could mention the supported joints here.

prashanthr05 commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,295 @@
+    xmlNodePtr child_xml = xmlNewChild(joint_xml, NULL, BAD_CAST "child", NULL);
+    xmlNewProp(child_xml, BAD_CAST "link", BAD_CAST model.getLinkName(childLink->getIndex()).c_str());
+
+    // TODO(traversaro) : handle joint limits and friction
```
This should be moved also in the documentation.

fjandrad replied:
I agree this is not mentioned in the description of the pr

fjandrad replied:
I think this is still pending

prashanthr05 commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,295 @@
+
+/**
+ * Add a <link> URDF element to the specified parent element with the specified link info.
+ */
```
Add the limitations that the current version does not support collisions and visuals.

traversaro replied:
Already documented in https://github.com/robotology/idyntree/pull/554/files#diff-05b37777f3462332b9cce4040d3f16deR64 , will add it there.

prashanthr05 commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,295 @@
+    xmlNewProp(robot, BAD_CAST "name", BAD_CAST "iDynTreeURDFModelExportModelName");
+    xmlDocSetRootElement(urdf_xml, robot);
+
+    // TODO(traversaro) : We are assuming that the model has no loops,
```
This must be mentioned in the documentation, explicitly. Since I notice there are users of iDynTree in the cable robots or parallel mechanisms community as well.

fjandrad replied:
I think this should be in the main readme of idyntree not only in this pr
I thought it was there but I could not find a mention to it now.

traversaro replied:
Related issue: https://github.com/robotology/idyntree/issues/555 .

fjandrad commented on src/model_io/urdf/tests/ModelExporterUnitTest.cpp:
```diff
@@ -0,0 +1,95 @@
+    ASSERT_EQUAL_DOUBLE(model.getNrOfLinks(), modelReloaded.getNrOfLinks());
+    ASSERT_EQUAL_DOUBLE(model.getNrOfJoints(), modelReloaded.getNrOfJoints());
+    ASSERT_EQUAL_DOUBLE(model.getNrOfDOFs(), modelReloaded.getNrOfDOFs());
+
```
Is there any special reason why don't you check also for the sensors?

traversaro replied:
Because at the moment they are not exported, so the test would fail. Relevant test will be added as soon as  sensors exporting is added.

prashanthr05 commented:
I think it will be helpful if we add an explicit print to console statement in the `URDFStringFromModel` or `URDFFromModel` functions about the elements that the user should not expect in the generated urdf file. Like the ones described in the PR. One step forward in avoiding any possible errors while using the generated urdf.

traversaro commented:
> I think it will be helpful if we add an explicit print to console statement in the `URDFStringFromModel` or `URDFFromModel` functions about the elements that the user should not expect in the generated urdf file. Like the ones described in the PR. One step forward in avoiding any possible errors while using the generated urdf.

I totally agree with this suggestion, but I expect to implement most of those features before the 0.12 release, so I would prefer to avoid overthinking how to report this problems to the user, as eventually we will need to remove them. If for the 0.12 release we still miss some features, then I would agree to add those error messages.

traversaro commented:
Reviews addressed in https://github.com/robotology/idyntree/pull/554/commits/15f4848c5c288006c475110911fea9cb50c1c99a .

fjandrad commented on src/model_io/urdf/src/URDFModelExport.cpp:
```diff
@@ -0,0 +1,345 @@
+    ok = ok && exportTransform(Transform(Rotation::Identity(), inertia.getCenterOfMass()), inertial);
+
+    xmlNodePtr inertia_xml = xmlNewChild(inertial, NULL, BAD_CAST "inertia", NULL);
+    RotationalInertiaRaw rotInertia = i.getRotationalInertiaWrtCenterOfMass();
```
I think this one was also meant to be `inertia `and not `i`

traversaro replied:
That happens when you modify the code and you do not try to recompile it. : ) 
Thanks for the check.

traversaro commented:
@fjandrad I am not sure why I cannot directly respond to your comments, anyhow I reply to them here. 

> I think this should be in the main readme of idyntree not only in this pr
I thought it was there but I could not find a mention to it now.

It make sense, I opened a new issue for this in https://github.com/robotology/idyntree/issues/555 .

> I think this is still pending

I already added it in https://github.com/robotology/idyntree/commit/15f4848c5c288006c475110911fea9cb50c1c99a#diff-05b37777f3462332b9cce4040d3f16deR69 for the main user facing documentation, and I added it in the specific function private docs in https://github.com/robotology/idyntree/pull/554/commits/ba04aaa212a5a5af602dda84a1968d860beef20c .

traversaro commented:
Friendly ping @prashanthr05 @fjandrad .

traversaro commented:
A legit problem in the build system on macOS was detected by Travis in https://travis-ci.org/robotology/idyntree/jobs/573683257, and  it should be fixed by https://github.com/robotology/idyntree/pull/554/commits/aaf1decbd6db8e3f847f4ee186ef1256672aa869 .

traversaro commented:
Friend ping @fjandrad

traversaro commented:
Thannks for the reviews @fjandrad and @prashanthr05 !
